packages/typespec-client-generator-core/test/decorators/legacy-hierarchy-building.test.ts

```diff
@@ -0,0 +1,535 @@
+  const smartKingSalmonModel = models.find((m) => m.name === "SmartKindSalmon");
+  const salmonPropertiesModel = models.find((m) => m.name === "SalmonProperties");
+  const kingSalmonPropertiesModel = models.find((m) => m.name === "KingSalmonProperties");
+  const smartKingSalmonPropertiesModel = models.find((m) => m.name === "SmartKindSalmonProperties");
+
+  ok(salmonModel);
+  ok(kingSalmonModel);
+  ok(smartKingSalmonModel);
+  ok(salmonPropertiesModel);
+  ok(kingSalmonPropertiesModel);
+  ok(smartKingSalmonPropertiesModel);
+
+  // SmartKindSalmon should inherit from KingSalmon instead of Salmon due to @Legacy.legacyHierarchyBuilding
+  strictEqual(smartKingSalmonModel.baseModel?.name, "KingSalmon");
+  strictEqual(kingSalmonModel.baseModel?.name, "Salmon");
+
+  // Verify discriminator property is correctly identified
+  strictEqual(salmonModel.discriminatorProperty?.name, "kind");
+  strictEqual(kingSalmonModel.discriminatorValue, "kingsalmon");
+  strictEqual(smartKingSalmonModel.discriminatorValue, "smartkingsalmon");
+
+  // Verify properties are correctly inherited
+  strictEqual(smartKingSalmonModel.properties.length, 1);
+  strictEqual(smartKingSalmonModel.properties[0].name, "kind");
+  strictEqual(smartKingSalmonPropertiesModel.properties.length, 2);
+  strictEqual(smartKingSalmonPropertiesModel.properties[0].name, "farmed");
+  strictEqual(smartKingSalmonPropertiesModel.properties[0].type.kind, "boolean");
+  strictEqual(smartKingSalmonPropertiesModel.properties[1].name, "size");
+  strictEqual(smartKingSalmonPropertiesModel.properties[1].type.kind, "boolean");
+
+  strictEqual(kingSalmonModel.properties.length, 2);
+  strictEqual(kingSalmonModel.properties[0].name, "kind");
+  strictEqual(kingSalmonModel.properties[1].type, kingSalmonPropertiesModel);
+  strictEqual(kingSalmonPropertiesModel.properties.length, 2);
+  strictEqual(kingSalmonPropertiesModel.properties[0].name, "farmed");
+  strictEqual(kingSalmonPropertiesModel.properties[0].type.kind, "constant");
+  strictEqual(kingSalmonPropertiesModel.properties[0].type.value, false);
+  strictEqual(kingSalmonPropertiesModel.properties[1].name, "size");
+  strictEqual(kingSalmonPropertiesModel.properties[1].type.kind, "int32");
+  strictEqual(salmonModel.properties.length, 2);
+  strictEqual(salmonModel.properties[0].name, "kind");
+  strictEqual(salmonModel.properties[0].type.kind, "string");
+  strictEqual(salmonModel.properties[1].type, salmonPropertiesModel);
+  // Verify .discriminatedSubtypes is correctly populated
+  ok(salmonModel.discriminatedSubtypes);
+  strictEqual(salmonModel.discriminatedSubtypes["kingsalmon"], kingSalmonModel);
+  strictEqual(salmonModel.discriminatedSubtypes["smartkingsalmon"], smartKingSalmonModel);
+  ok(kingSalmonModel.discriminatedSubtypes);
+  strictEqual(kingSalmonModel.discriminatedSubtypes["smartkingsalmon"], smartKingSalmonModel);
+  ok(!smartKingSalmonModel.discriminatedSubtypes);
+});
+
+it("circular inheritance", async () => {
```

Comment:
Should we really allow this case? First only B and A don't have the same root parent, how do we know the discrinminator property? A and B have circular inheritance, this is not allowed in typespec yet. Another inference is if A and B have circular relationship which means they should be the same, why we define this thing?

<img width="1563" height="707" alt="image" src="https://github.com/user-attachments/assets/22000d1e-f2e7-4de0-a7d8-1a38753b1a0d" />